swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/services/ContractsController.java

```diff
@@ -88,18 +93,44 @@ public Double getContractCoverage(BillableUsage usage) throws ContractMissingExc
       throw new ContractMissingException(
           String.format("No contract info found for usage! %s", usage));
     }
-    Integer totalUnderContract =
-        contracts.stream()
-            .filter(contract -> isValidContract(contract, usage))
-            .map(
-                c ->
-                    c.getMetrics().stream()
-                        .filter(metric -> metric.getMetricId().equals(contractMetricId))
-                        .map(Metric::getValue)
-                        .reduce(0, Integer::sum))
-            .reduce(0, Integer::sum);
-    log.debug("Total contract coverage is {} for usage {} ", totalUnderContract, usage);
-    return Double.valueOf(totalUnderContract);
+
+    boolean hasGratisStatus =
+        SubscriptionDefinition.isMetricGratis(usage.getProductId(), usage.getMetricId());
+    ContractCoverage coverage = new ContractCoverage(contractMetricId);
+    for (Contract contract : contracts) {
+      if (isValidContract(contract, usage)) {
+        var value =
+            contract.getMetrics().stream()
+                .filter(metric -> metric.getMetricId().equals(contractMetricId))
+                .map(Metric::getValue)
+                .reduce(0, Integer::sum);
+        if (hasGratisStatus && isGratisContract(contract)) {
+          // Directly return that the coverage is gratis.
+          coverage.setGratis(true);
+          return coverage;
+        } else {
+          coverage.add(value);
+        }
+      }
+    }
+
+    log.debug("Total contract coverage is {} for usage {} ", coverage, usage);
+    return coverage;
+  }
+
+  /**
+   * Check whether contract start and end dates applies the conditions for a gratis usage. From <a
+   * href="https://issues.redhat.com/browse/SWATCH-2571">SWATCH-2571</a>, the conditions are: -
+   * contract starts on the current month - and contracts does not end in the middle of the current
+   * month
+   */
+  private boolean isGratisContract(Contract contract) {
+    OffsetDateTime startDate = contract.getStartDate();
+    OffsetDateTime endDate = contract.getEndDate();
+    return startDate != null
+        && (startDate.isEqual(clock.startOfCurrentMonth())
+            || startDate.isAfter(clock.startOfCurrentMonth()))
+        && (endDate == null || !endDate.isBefore(clock.endOfCurrentMonth()));
```

Comment:
We simplified the logic during refinement, the only condition now is that the startDate is after the start of the current month:
```suggestion
    OffsetDateTime startDate = contract.getStartDate();
    return startDate != null && startDate.isAfter(clock.startOfCurrentMonth());
```

Reply:
Thanks for clarifying. Will change it.